### PR TITLE
Support HotSpotDiagnosticMXBean in OpenJ9 JDK21+

### PR DESCRIPTION
**Pull request by babsingh**
This will allow us to enable a JTReg test for Structured Concurrency.

JTReg test: StructuredTaskScope/StructuredThreadDumpTest

Related: #17634

### CODE REVIEW

**Comment by babsingh** [comment]
@pshipton @tajila I verified that `StructuredThreadDumpTest` passes with this change. Also, I didn't observe any side-effects of this change in personal builds. Can one of you please review these changes?

File: jcl/src/java.management/share/classes/com/ibm/java/lang/management/internal/ManagementUtils.java [hunk 1]
```diff
@@ -774,6 +778,12 @@ static void registerAll() {
 				.addInterface(openj9.lang.management.OpenJ9DiagnosticsMXBean.class)
 				.validateAndRegister();
 
+			/*[IF JAVA_SPEC_VERSION >= 21]*/
+			create(HOTSPOT_DIAGNOSTICS_MXBEAN_NAME, com.sun.management.internal.ExtendedHotSpotDiagnostic.getInstance())
```
Comment (pshipton) [inline jcl/src/java.management/share/classes/com/ibm/java/lang/management/internal/ManagementUtils.java:782]: I think this needs a module check, like you can find in PlatformMBeanProvider.java. ExtendedHotSpotDiagnostic can only be created if the jdk.management module is loaded.

**Comment by tajila** [comment]
Does the use of HotSpotDiagnosticMXBean require `-XX:+UnlockDiagnosticVMOptions ` on hotspot?

**Comment by babsingh** [comment]
> Does the use of HotSpotDiagnosticMXBean require `-XX:+UnlockDiagnosticVMOptions ` on hotspot?

I don't believe that it requires `-XX:+UnlockDiagnosticVMOptions`. [StructuredThreadDumpTest](https://github.com/ibmruntimes/openj9-openjdk-jdk21/blob/5b63a9eb7bc4bb8bf47f8e3a1e50c87c1e553cda/test/jdk/java/util/concurrent/StructuredTaskScope/StructuredThreadDumpTest.java#L24-L31) uses `HotSpotDiagnosticMXBean` without `-XX:+UnlockDiagnosticVMOptions`.

File: jcl/src/java.management/share/classes/com/ibm/java/lang/management/internal/ManagementUtils.java (hunk 1 above)
Comment (babsingh) [reply jcl/src/java.management/share/classes/com/ibm/java/lang/management/internal/ManagementUtils.java]: Not completely sure. Let's try removing the changes from `ManagementUtils`. If testing shows side-effects, then I will add it back in `ManagementUtils`. In `PlatformMBeanProvider`, I have added a `jdk.management` module check for `HotSpotDiagnosticMXBean`.

**Comment by babsingh** [comment]
All feedback comments have been addressed.

**Comment by babsingh** [comment]
Previous feedback has been addressed: removed the module check and added error handling as per the Javadoc.

**Comment by pshipton** [comment]
What test suite(s) did you run already?

File: jcl/src/jdk.management/share/classes/com/sun/management/internal/ExtendedHotSpotDiagnostic.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	public void dumpHeap(String outputFile, boolean live) throws IOException {
+		throw new UnsupportedOperationException();
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public List<VMOption> getDiagnosticOptions() {
+		return null;
```
Comment (pshipton) [inline jcl/src/jdk.management/share/classes/com/sun/management/internal/ExtendedHotSpotDiagnostic.java:68]: I don't see null as an option in the javadoc,  it needs to be an empty list.

File: jcl/src/jdk.management/share/classes/com/sun/management/internal/ExtendedHotSpotDiagnostic.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public VMOption getVMOption(String name) {
+		if (name == null) {
+			throw new NullPointerException("A non-null name is required");
+		}
+		throw new IllegalArgumentException("VM option of the given name does not exist: " + name);
```
Comment (pshipton) [inline jcl/src/jdk.management/share/classes/com/sun/management/internal/ExtendedHotSpotDiagnostic.java:79]: The message could explain that OpenJ9 doesn't support the API. i.e. the option may exist, but OpenJ9 is not going to be retrieving the value.

**Comment by babsingh** [comment]
> What test suite(s) did you run already?

JDK21: `sanity.functional,sanity.openjdk,extended.openjdk`. JDK21 personal build: https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/17143/

JDK17: `sanity.functional`. JDK17 personal build: https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/17144/

Also, accounted for the recent feedback: an empty list is returned and exception messages have been updated.

**Comment by pshipton** [comment]
jenkins test extended amac jdk11,jdk21

File: jcl/src/jdk.management/share/classes/com/sun/management/internal/ExtendedHotSpotDiagnostic.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	public void dumpHeap(String outputFile, boolean live) throws IOException {
+		throw new UnsupportedOperationException();
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public List<VMOption> getDiagnosticOptions() {
+		return new ArrayList<>();
```
Comment (pshipton) [inline jcl/src/jdk.management/share/classes/com/sun/management/internal/ExtendedHotSpotDiagnostic.java:69]: Pls use Collection.EMPTY_LIST

Comment (babsingh) [reply jcl/src/jdk.management/share/classes/com/sun/management/internal/ExtendedHotSpotDiagnostic.java]: Updated.

**Comment by babsingh** [comment]
Old PR builds: https://openj9-jenkins.osuosl.org/job/PullRequest-OpenJ9/4025/

**Review by pshipton — APPROVED** [review]

File: jcl/src/java.management/share/classes/com/ibm/java/lang/management/internal/ManagementUtils.java [hunk 5]
```diff
@@ -770,10 +770,6 @@ static void registerAll() {
 				.addInterface(com.ibm.lang.management.JvmCpuMonitorMXBean.class)
 				.validateAndRegister();
 
-			create(OPENJ9_DIAGNOSTICS_MXBEAN_NAME, openj9.lang.management.internal.OpenJ9DiagnosticsMXBeanImpl.getInstance())
-				.addInterface(openj9.lang.management.OpenJ9DiagnosticsMXBean.class)
-				.validateAndRegister();
```
Comment (keithc-ca) [inline jcl/src/java.management/share/classes/com/ibm/java/lang/management/internal/ManagementUtils.java:775]: Please help me understand why this was removed. This code is specific to Java 8 where there's no possibility that  `OpenJ9DiagnosticsMXBeanImpl.getInstance()` might return `null` (even though the code would properly handle that situation). Why do we no longer want this bean to be registered?

Comment (pshipton) [reply jcl/src/java.management/share/classes/com/ibm/java/lang/management/internal/ManagementUtils.java:775]: Ah good catch, I missed it was specific to 8 (and we didn't test 8). So we need to add it back and add the HotspotBean there too.

Comment (babsingh) [reply jcl/src/java.management/share/classes/com/ibm/java/lang/management/internal/ManagementUtils.java:775]: Opened https://github.com/eclipse-openj9/openj9/pull/17726 to readd the registration calls.

**Pull request merged**